File: leetcode/c++/LCR_142_he_bing_liang_ge_pai_xu_de_lian_biao_lcof.cpp
// https://leetcode.cn/problems/he-bing-liang-ge-pai-xu-de-lian-biao-lcof/description

#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* trainningPlan(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode(0);
        ListNode* tmp = head;
        while (l1 != nullptr && l2 != nullptr) {
            if (l1->val <= l2->val) {
                tmp->next = l1;
                l1 = l1->next;
            } else {
                tmp->next = l2;
                l2 = l2->next;
            }
            tmp = tmp->next;
        }
        tmp->next = l1 != nullptr ? l1 : l2;
        return head->next;
    }
};

// O(n+m)， O(1)
// 复习x

File: leetcode/golang/LCR_142_he_bing_liang_ge_pai_xu_de_lian_biao_lcof.go
type ListNode struct {
    Val int
    Next *ListNode
}

func trainningPlan(l1 *ListNode, l2 *ListNode) *ListNode {
    head := &ListNode {
        Val:0,
        Next:nil,
    }
    tmp := head
    for l1 != nil && l2 != nil {
        if l1.Val <= l2.Val {
            tmp.Next = l1
            l1 = l1.Next
        } else {
            tmp.Next = l2
            l2 = l2.Next
        }
        tmp = tmp.Next
    }
    if l1 != nil {
        tmp.Next = l1
    } 
    if l2 != nil {
        tmp.Next = l2
    }
    return head.Next
}